pkg/agent/cniserver/ipam/antrea_ipam_controller.go

```diff
@@ -81,21 +117,102 @@ func (c *AntreaIPAMController) Run(stopCh <-chan struct{}) {
 	}()
 
 	klog.InfoS("Starting", "controller", controllerName)
-	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced, c.ipPoolInformer.Informer().HasSynced, c.podInformer.HasSynced) {
 		return
 	}
 
 	<-stopCh
 }
 
-func (c *AntreaIPAMController) getIPPoolsByNamespace(namespace string) []string {
+func (c *AntreaIPAMController) getIPPoolsByPod(namespace, name string) ([]string, []net.IP, *crdv1a2.IPAddressOwner, error) {
+	// Find IPPool by Pod
+	var IPs []net.IP
+	var reservedOwner *crdv1a2.IPAddressOwner
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		// For CNI DEL case, getting Pod may fail. Try to get information from allocated IPs of all IPPools.
+		klog.Infof("Get pod %s/%s failed, err=%+v", namespace, name, err)
+		ipPools, _ := c.ipPoolInformer.Informer().GetIndexer().ByIndex(podIndex, k8s.NamespacedName(namespace, name))
+		for _, item := range ipPools {
+			ipPool := item.(*crdv1a2.IPPool)
+			if ipPool.Spec.IPVersion != 4 {
+				continue
+			}
+			for _, IPAddress := range ipPool.Status.IPAddresses {
+				if IPAddress.Owner.Pod != nil && IPAddress.Owner.Pod.Namespace == namespace && IPAddress.Owner.Pod.Name == name {
```

Comment:
Could you please clarify - is this code for re-allocation of released Pod? I think with current code Pod Owner will be released from state.

Reply:
No, this is to release the allocation.
Only CNI DEL case will enter this code snippet, the PodOwner will be removed at last.

Reply:
Got it, thanks. But I think we don't need `reservedOwner` and addresses for CNI Del? The only thing needed is the poolName?

Reply:
I would suggest to move the code under L132 to a separate routine (that would only find the pool), and invoke it only from CNI Del. I think this will make the code more readible.

Reply:
Yes. Removed reservedOwner in this code snippet. Thanks.

Reply:
> I would suggest to move the code under L132 to a separate routine (that would only find the pool), and invoke it only from CNI Del. I think this will make the code more readible.

I think the code can be shared with CNI Add/Check. Do you think we can passthrough isDel mark to this function and check isDel at L132?

Reply:
I think code that deals with Del scenario only (L132-146) can be wrapped in it own function and called from CNI Del directly, since it has no use with Add/Check.

Reply:
The call of this function has 3 levels of nesting: `owns`->`getPoolAllocatorByPod`->`getIPPoolsByPod`
Thus it is not easy to wrap this code into a separate function and call directly since we have a lot of shared logic in `owns` and `getPoolAllocatorByPod`